File: business-logic/classic-scanlog-core/tests/semantic_conformance/dds_header.rs
//! Input-only public DDS byte parser and validation observations.

use super::{RunnerResult, invalid};
use classic_file_io_core::dds::DDSHeader;
use serde_json::{Value, json};

/// Parses raw input bytes and projects the native fields and validation methods.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    let bytes = fixture["bytes"]
        .as_array()
        .ok_or_else(|| invalid("DDS bytes must be an array"))?
        .iter()
        .map(|value| {
            value
                .as_u64()
                .and_then(|value| u8::try_from(value).ok())
                .ok_or_else(|| invalid("DDS input byte is invalid"))
        })
        .collect::<Result<Vec<_>, _>>()?;
    if matches!(fixture["operation"].as_str(), Some("files" | "validate")) {
        let temporary = tempfile::tempdir()?;
        let path = temporary.path().join("texture.dds");
        let missing = temporary.path().join("missing.dds");
        std::fs::write(&path, &bytes)?;
        if fixture["operation"] == "files" {
            use classic_file_io_core::{FileIOCore, FileIOError};
            let io = FileIOCore::default();
            let runtime = classic_shared_core::get_runtime();
            let dimensions = runtime
                .block_on(io.read_dds_header(&path))?
                .map(|header| [header.width, header.height]);
            let batch = io
                .read_dds_headers_batch(vec![path.clone(), missing.clone()])
                .into_iter()
                .map(|(path, header)| {
                    json!([
                        path.file_name().unwrap().to_string_lossy(),
                        header.map(|header| [header.width, header.height])
                    ])
                })
                .collect::<Vec<_>>();
            let missing_error = match runtime.block_on(io.read_dds_header(&missing)) {
                Err(FileIOError::IoError(_)) => Some("io_error"),
                Err(error) => return Err(error.into()),
                Ok(_) => None,
            };
            return Ok(
                json!({"dimensions":dimensions,"batch":batch,"missingError":missing_error,"bytes":std::fs::read(path)?}),
            );
        }
        use classic_file_io_core::dds::{DDSAnalyzer, GameTarget};
        let analyzer = DDSAnalyzer::new(GameTarget::Fallout4);
        let issues = analyzer
            .validate_file(&path)
            .into_iter()
            .map(|issue| issue.message)
            .collect::<Vec<_>>();
        let batch = analyzer
            .validate_batch(&[path.clone(), missing])
            .into_iter()
            .map(|(path, issues)| {
                json!([
                    path.file_name().unwrap().to_string_lossy(),
                    issues
                        .into_iter()
                        .map(|issue| issue.message)
                        .collect::<Vec<_>>()
                ])
            })
            .collect::<Vec<_>>();
        let dimensions = fixture["dimensions"]
            .as_array()
            .ok_or_else(|| invalid("DDS dimensions must be an array"))?;
        let width = u32::try_from(
            dimensions[0]
                .as_u64()
                .ok_or_else(|| invalid("invalid width"))?,
        )?;
        let height = u32::try_from(
            dimensions[1]
                .as_u64()
                .ok_or_else(|| invalid("invalid height"))?,
        )?;
        return Ok(
            json!({"issues":issues,"batch":batch,"dimensionIssues":DDSAnalyzer::validate_dimensions(width,height).into_iter().map(|issue|issue.message).collect::<Vec<_>>(),"bytes":std::fs::read(path)?}),
        );
    }
    let header = DDSHeader::from_bytes(&bytes).map_err(|error| invalid(&error.to_string()))?;
    Ok(json!({"bytes":bytes, "header":header.map(|header| json!({
        "width":header.width,"height":header.height,"depth":header.depth,
        "mipmaps":header.mipmap_count,"format":header.format,
        "powerOfTwo":header.has_power_of_2_dimensions(),
        "validBcDimensions":header.has_valid_bc_dimensions(),
        "reasonableSize":header.is_reasonable_size(),
        "hasMipmaps":header.has_mipmaps(),"bcCompressed":header.is_bc_compressed()
    }))}))
}

File: tools/binding_compliance/conformance/families/dds_header.py
"""Executable DDS byte parsing and complete public header validation facts."""

from collections.abc import Mapping
from functools import partial
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy


def _observed(parsed: bool, observation: Mapping[str, Any]) -> bool:
    """Require native header fields and every query result to agree with raw bytes."""
    if set(observation) != {"bytes", "header"}:
        return False
    raw = observation["bytes"]
    if not isinstance(raw, list) or not all(
        type(value) is int and 0 <= value <= 255 for value in raw
    ):
        return False
    header = observation["header"]
    if not parsed:
        return header is None and (len(raw) < 128 or raw[:4] != [68, 68, 83, 32])
    if len(raw) < 128 or raw[:4] != [68, 68, 83, 32] or not isinstance(header, Mapping):
        return False
    if set(header) != {
        "width",
        "height",
        "depth",
        "mipmaps",
        "format",
        "powerOfTwo",
        "validBcDimensions",
        "reasonableSize",
        "hasMipmaps",
        "bcCompressed",
    }:
        return False
    width, height, mipmaps = (
        int.from_bytes(bytes(raw[offset : offset + 4]), "little")
        for offset in (16, 12, 28)
    )
    if any(
        type(header[key]) is not int for key in ("width", "height", "depth", "mipmaps")
    ):
        return False
    return (
        header["width"] == width
        and header["height"] == height
        and header["depth"] == 1
        and header["mipmaps"] == max(1, mipmaps)
        and header["format"] == "BC3_UNorm_sRGB"
        and header["powerOfTwo"]
        is (
            width > 0
            and height > 0
            and width & (width - 1) == 0
            and height & (height - 1) == 0
        )
        and header["validBcDimensions"] is (width % 4 == 0 and height % 4 == 0)
        and header["reasonableSize"] is (0 < width <= 16384 and 0 < height <= 16384)
        and header["hasMipmaps"] is (mipmaps > 1)
        and header["bcCompressed"] is True
    )


def _file_observed(observation: Mapping[str, Any]) -> bool:
    """Require single/batch dimensions, preserved raw bytes and a typed missing-file failure."""
    if set(observation) != {"dimensions", "batch", "missingError", "bytes"}:
        return False
    dimensions = observation["dimensions"]
    return (
        (
            dimensions is None
            or isinstance(dimensions, list)
            and len(dimensions) == 2
            and all(type(value) is int and value >= 0 for value in dimensions)
        )
        and observation["batch"] == [["texture.dds", dimensions], ["missing.dds", None]]
        and observation["missingError"] == "io_error"
        and isinstance(observation["bytes"], list)
        and all(
            type(value) is int and 0 <= value <= 255 for value in observation["bytes"]
        )
    )


def _validation(observation: Mapping[str, Any]) -> bool:
    """Require actual issue strings and batch filtering to agree with single-file validation."""
    if set(observation) != {"issues", "batch", "dimensionIssues", "bytes"}:
        return False
    return (
        all(
            isinstance(observation[key], list)
            and all(
                isinstance(message, str) and message for message in observation[key]
            )
            for key in ("issues", "dimensionIssues")
        )
        and observation["batch"]
        == ([["texture.dds", observation["issues"]]] if observation["issues"] else [])
        + [["missing.dds", ["Unable to read DDS file"]]]
        and isinstance(observation["bytes"], list)
        and all(
            type(value) is int and 0 <= value <= 255 for value in observation["bytes"]
        )
    )


DDS_HEADER_COVERAGE_POLICY = FamilyCoveragePolicy(
    "dds-header",
    tuple(
        CoveragePredicate(
            id="parsed-header" if parsed else "rejected-header",
            capability_id="dds-header.parse",
            action="dds-header.parse",
            observation_family="header",
            rust_symbols=("DDSHeader",),
            matches=partial(_observed, parsed),
            runtime_operations=(
                None,
                "from_bytes",
                "has_power_of_2_dimensions",
                "has_valid_bc_dimensions",
                "is_reasonable_size",
                "has_mipmaps",
                "is_bc_compressed",
            )
            if parsed
            else ("from_bytes",),
        )
        for parsed in (True, False)
    )
    + (
        CoveragePredicate(
            "file-readers",
            "dds-header.files",
            "dds-header.files",
            "file-headers",
            ("FileIOCore",),
            _file_observed,
            runtime_operations=("read_dds_header", "read_dds_headers_batch"),
        ),
        CoveragePredicate(
            "dds-validation",
            "dds-header.validate",
            "dds-header.validate",
            "dds-issues",
            ("DDSAnalyzer",),
            _validation,
            runtime_operations=(None,),
        ),
    ),
)
